lib/active_merchant/billing/gateways/pay_arc.rb

```diff
@@ -0,0 +1,345 @@
+        post['card_number'] = creditcard.number
+        post['exp_month'] = creditcard.month
+        post['exp_year'] = creditcard.year
+        post['cvv'] = creditcard.verification_value unless creditcard.verification_value.nil?
+        post['card_holder_name'] = options[:card_holder_name] || "#{creditcard.first_name} #{creditcard.last_name}"
+      end
+
+      def add_address(post, creditcard, options)
+        post['address_line1'] = options[:address_line1]
+        post['address_line2'] = options[:address_line2]
+        post['city'] = options[:city]
+        post['state'] = options[:state]
+        post['zip'] = options[:zip]
+        post['country'] = options[:country]
+      end
+
+      def add_money(post, money, options)
+        post['amount'] = money
+        post['currency'] = (options[:currency] || currency(money))
+        post['statement_description'] = options[:statement_description]
+      end
+
+      def headers(api_key)
+        {
+          'Authorization' => 'Bearer ' + api_key.strip,
+          'Accept' => 'application/json',
+          'User-Agent' => "PayArc ActiveMerchantBindings/#{ActiveMerchant::VERSION}"
+        }
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      rescue JSON::ParserError
+        body
+      end
+
+      def commit(action, parameters)
+        url = (test? ? test_url : live_url)
+        headers = headers(@options[:api_key])
+        end_point = "#{url}/#{action}"
+        begin
+          response = ssl_post(end_point, post_data(action, parameters), headers)
+          parsed_response = parse(response)
+
+          Response.new(
+            success_from(parsed_response, action),
+            message_from(parsed_response, action),
+            parsed_response,
+            test: test?,
+            authorization: parse_response_id(parsed_response)
+          )
+        rescue ResponseError => e
+          parsed_response = parse(e.response.body)
+          Response.new(
+            false,
+            message_from(parsed_response, action),
+            parsed_response,
+            test: test?,
+            authorization: nil
+          )
+        end
+      end
+
+      def success_from(response, action)
+        if action == STANDARD_ACTIONS[:token]
+          token = parse_response_id(response)
+          (!token.nil? && !token.empty?)
+        elsif response
+          return SUCCESS_STATUS.include? response['data']['status'] if response['data']
+        end
+      end
+
+      def message_from(response, action)
+        if success_from(response, action)
+          if action == STANDARD_ACTIONS[:token]
+            return response['data']['id']
+          else
+            return response['data']['status']
+          end
+        else
+          return response['message']
+        end
+      end
+
+      def parse_response_id(response)
+        response['data']['id'] if response && response['data']
+      end
+
+      def post_data(action, params)
```

Comment:
`action` is not being used here and can be removed

Reply:
Hi @jessiagee 

Added this and pushed.
Also, we have added a filter to pass gateway-specific fields to PayArc.

Reply:
Hi @jessiagee,

We have also added, credit method as well now and pushed to same branch

Reply:
Hi @jessiagee,

One small clarification needed, the ActiveMerchant refund method needs money as a mandatory argument. But when we do full refund we can do it without sending money to the PAYARC API. 

Will there a scenario in Spreedly where you will call refund method without money? Should nI make the amount an optional parameter in active metchant ? 

Please advice.

Senthil

Reply:
Hi Senthil,

Here's the Spreedly API for refund method - https://docs.spreedly.com/reference/api/v1/#credit-refund

[Full amounts](https://docs.spreedly.com/reference/api/v1/#full-amount75) don't require an `amount` passed in the transaction body while [partial amounts](https://docs.spreedly.com/reference/api/v1/#partial-amount76) do require an `amount` passed in the transaction body.

The `money` parameter should stay required, and Spreedly will handle `full` vs. `partial` based off of the body of the REST request.

Let me know if you have any other questions!

Jessi